mirrord/layer/src/file/ops.rs

```diff
@@ -217,6 +219,98 @@ pub(crate) fn fdopen(fd: RawFd, rawish_mode: Option<&CStr>) -> Detour<*mut FILE>
     Detour::Success(result)
 }
 
+/// creates a directory stream for the `remote_fd` in the agent
+#[tracing::instrument(level = "trace")]
+pub(crate) fn fdopendir(fd: RawFd) -> Detour<usize> {
+    // usize == ptr size
+    // we don't return a pointer to an address that contains DIR
+
+    let open_files = OPEN_FILES.lock()?;
+    let remote_file_fd = open_files.get(&fd).ok_or(Bypass::LocalFdNotFound(fd))?;
+
+    let (dir_channel_tx, dir_channel_rx) = oneshot::channel();
+
+    let open_dir_request = OpenDir {
+        remote_fd: *remote_file_fd,
+        dir_channel_tx,
+    };
+
+    blocking_send_file_message(HookMessageFile::OpenDir(open_dir_request))?;
+
+    let OpenDirResponse { fd } = dir_channel_rx.blocking_recv()??;
+
+    let fake_local_dir_name = CString::new(fd.to_string())?;
+    let local_dir_fd = unsafe { create_local_fake_file(fake_local_dir_name, *remote_file_fd) }?;
+    OPEN_DIRS.lock()?.insert(local_dir_fd, fd);
+
+    Detour::Success(local_dir_fd as usize)
+}
+
+// fetches the current entry in the directory stream created by `fdopendir`
+#[tracing::instrument(level = "trace")]
+pub(crate) fn readdir_r(
+    dirp: *mut DIR,
+    entry: *mut dirent,
+    result: *mut *mut dirent,
+) -> Detour<c_int> {
+    trace!("readdir_r -> {dirp:#?} {entry:#?} {result:#?}");
+    let remote_fd = OPEN_DIRS.lock()?.get(&(dirp as i32)).cloned().unwrap();
+
+    let (dir_channel_tx, dir_channel_rx) = oneshot::channel();
+
+    let requesting_dir = ReadDir {
+        remote_fd,
+        dir_channel_tx,
+    };
+
+    blocking_send_file_message(HookMessageFile::ReadDir(requesting_dir))?;
+
+    let ReadDirResponse { direntry } = dir_channel_rx.blocking_recv()??;
+
+    if let Some(direntry) = direntry {
+        let direntry_name = direntry.name.as_slice();
+
+        #[cfg(target_os = "linux")]
+        let mut entry_name: [i8; 256] = [0; 256];
+
+        #[cfg(target_os = "macos")]
+        let mut entry_name: [i8; 1024] = [0; 1024];
+
+        let casted_name: Vec<i8> = direntry_name.iter().map(|c| *c as i8).collect();
```

Comment:
Cast the entry name to `Vec<u8>` then write it to the `entry.d_name` using `std::ptr::copy`
See opt 2 here (not exact, but similar)
https://dev.to/kgrech/7-ways-to-pass-a-string-between-rust-and-c-4ieb